src/main/java/org/privacyidea/JSONParser.java

```diff
@@ -0,0 +1,345 @@
+import static org.privacyidea.PIConstants.USERNAME;
+import static org.privacyidea.PIConstants.VALUE;
+import static org.privacyidea.PIConstants.VERSION_NUMBER;
+import static org.privacyidea.PIConstants.WEBAUTHN_SIGN_REQUEST;
+
+public class JSONParser {
+
+    private final PrivacyIDEA privacyIDEA;
+
+    public JSONParser(PrivacyIDEA privacyIDEA) {
+        this.privacyIDEA = privacyIDEA;
+    }
+
+    public String formatJson(String json) {
+        if (json == null || json.isEmpty()) return "";
+
+        JsonObject obj;
+        Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        try {
+            obj = JsonParser.parseString(json).getAsJsonObject();
+        } catch (JsonSyntaxException e) {
+            privacyIDEA.error(e);
+            return json;
+        }
+
+        return gson.toJson(obj);
+    }
+
+    String parseAuthToken(String serverResponse) {
+        if (serverResponse != null && !serverResponse.isEmpty()) {
+            JsonElement root = JsonParser.parseString(serverResponse);
+            if (root != null) {
+                try {
+                    JsonObject obj = root.getAsJsonObject();
+                    return obj.getAsJsonObject(RESULT).getAsJsonObject(VALUE).getAsJsonPrimitive(TOKEN).getAsString();
+                } catch (Exception e) {
+                    privacyIDEA.error("Response did not contain an authorization token: " + formatJson(serverResponse));
+                }
+            }
+        } else {
+            privacyIDEA.error("/auth response was empty or null!");
+        }
+        return "";
+    }
+
+    PIResponse parsePIResponse(String serverResponse) {
+        PIResponse response = new PIResponse();
+        if (serverResponse == null) return response;
+        response.rawMessage = serverResponse;
+
+        if (serverResponse.isEmpty()) return response;
+
+        JsonObject obj;
+        try {
+            obj = JsonParser.parseString(serverResponse).getAsJsonObject();
+        } catch (JsonSyntaxException e) {
+            privacyIDEA.error(e);
+            return response;
+        }
+
+        response.id = getString(obj, ID);
+        response.piVersion = getString(obj, VERSION_NUMBER);
+        response.signature = getString(obj, SIGNATURE);
+        response.jsonRPCVersion = getString(obj, JSONRPC);
+
+        JsonObject result = obj.getAsJsonObject(RESULT);
+        if (result != null) {
+            response.status = getBoolean(result, STATUS);
+            response.value = getBoolean(result, VALUE);
+
+            JsonElement errElem = result.get(ERROR);
+            if (errElem != null && !errElem.isJsonNull()) {
+                JsonObject errObj = result.getAsJsonObject(ERROR);
+                response.error = new PIResponse.Error();
+                response.error.code = getInt(errObj, CODE);
+                response.error.message = getString(errObj, MESSAGE);
+            }
+        }
+
+        JsonElement detailElem = obj.get(DETAIL);
```

Comment:
Isn't there a more generic way to parse this?

The "detail" in the section could also contain user information and other stuff, that might be needed by certain plugins.
But it looks like as if this would not be contained in the ``response`` object which you return later.

Reply:
The PIResonpse object contains the original json aswell, so if there is need to look for a special field in that, it could be done.
Since json parsing is not that elegant and pretty lengthy in java, the response object was added. However, parsing the response of the server into a map for example would result in the plugin doing:
`if (map.contains(SOME_KEY)) 
if(map.get(SOME_KEY) != null)) do something
else use some default value`
over and over again. The response object provides easier access and non-null default values so that no runtime exception occur from this.
If we can have a contract between the plugins and server (i.e. a class definition with fixed member variables) for the responses, we could establish that and the parsing can be done automatically by gson.